CryptoAnalysis/src/main/java/crypto/providerdetection/ProviderDetection.java

```diff
@@ -0,0 +1,377 @@
+							JAssignStmt stmt = (JAssignStmt) unit;
+							Value rightVal = stmt.getRightOp();
+							if (rightVal instanceof JStaticInvokeExpr) {
+								JStaticInvokeExpr exp = (JStaticInvokeExpr) rightVal;
+									
+								SootMethod method = exp.getMethod();
+								String methodName = method.getName();
+									
+								SootClass methodRef = method.getDeclaringClass();
+								String refName = methodRef.toString();
+//								System.out.println(refName);
+									
+								int parameterCount = method.getParameterCount();
+									
+								//list of JCA engine classes that are supported as CryptSL rules
+								String[] crySLRules = new String[] {"java.security.SecureRandom", "java.security.MessageDigest",
+																	"java.security.Signature", "javax.crypto.Cipher", 
+																	"javax.crypto.Mac", "javax.crypto.SecretKeyFactory",
+																	"javax.crypto.KeyGenerator", "java.security.KeyPairGenerator",
+																	"java.security.AlgorithmParameters", "java.security.KeyStore",
+																	"javax.net.ssl.KeyManagerFactory", "javax.net.ssl.SSLContext",
+																	"javax.net.ssl.TrustManagerFactory"};
+									
+								boolean ruleFound = Arrays.asList(crySLRules).contains(refName);
+									
+								if((ruleFound) && (methodName.matches("getInstance")) && (parameterCount==2) ) {
+									Value vl = exp.getArg(1);
+//									System.out.println(vl.toString());
+									String strType = getProviderType(vl);
+//									System.out.println("The provider used is of type: "+strType);
+										
+									if(strType.matches("java.security.Provider")) {
+										this.provider = getProvider(stmt, sootMethod, vl, icfg);
+										if(ruleExists(provider, refName)) {
+											rules = chooseRules(rules, provider, refName);
+											break outerloop;
+										} 
+									}
+										
+									else if (strType.matches("java.lang.String")) {
+//										for(Unit u : body.getUnits()) {
+//											if(u instanceof JIfStmt) {
+//												JIfStmt l = (JIfStmt) u;
+//												System.out.println("labeled: "+l.toString() + " "+l.getClass());
+//												System.out.println(l.getUnitBoxes());
+//											}
+//										}
+										
+										this.provider = getProviderWhenTypeString(vl, body);
+										if(ruleExists(provider, refName)) {
+											rules = chooseRules(rules, provider, refName);
+											break outerloop;
+										}
+									}
+								}
+							}
+						}
+					}
+				}	
+			}
+		}
+	 			
+		return rules;
+	}
+	
+	
+	
+	//methods used from the "doAnalysis() method"
+	//-----------------------------------------------------------------------------------------------------------------
+	
+	private String getProviderType(Value vl) {
+		Type type = vl.getType();
+//		System.out.println("Soot type: "+type);
+		String strType = type.toString();
+		return strType;
+	}
+	
+	
+	private String getProviderWhenTypeString(Value vl, Body body) {
```

Comment:
@kruegers This method does provider detection when the provider is of type `java.lang.String`.